docs/HowToGuides/GettingStarted.md

```diff
@@ -0,0 +1,609 @@
+# How to Set Up an Edit-Build-Test-Debug Loop
+
+This document describes how to set up a development loop for people interested
+in contributing to Swift.
+
+If you are only interested in building the
+toolchain as a one-off, there are a couple of differences:
+1. You can ignore the parts related to Sccache.
+2. You can stop reading after
+   [Building the project for the first time](#building-the-project-for-the-first-time).
+
+## Table of Contents
+
+- [System Requirements](#system-requirements)
+- [Cloning the project](#cloning-the-project)
+  - [Troubleshooting cloning issues](#troubleshooting-cloning-issues)
+- [Installing dependencies](#installing-dependencies)
+  - [macOS](#macOS)
+  - [Ubuntu Linux](#ubuntu-linux)
+- [Building the project for the first time](#building-the-project-for-the-first-time)
+  - [Spot check dependencies](#spot-check-dependencies)
+  - [Understanding the pieces](#understanding-the-pieces)
+  - [The actual build](#the-actual-build)
+  - [Troubleshooting build issues](#troubleshooting-build-issues)
+- [Editing code](#editing-code)
+  - [Setting up your fork](#setting-up-your-fork)
+  - [First time Xcode setup](#first-time-xcode-setup)
+  - [Editing](#editing)
+  - [Incremental builds with Ninja](#incremental-builds-with-ninja)
+  - [Incremental builds with Xcode](#incremental-builds-with-xcode)
+  - [Spot checking an incremental build](#spot-checking-an-incremental-build)
+- [Reproducing an issue](#reproducing-an-issue)
+- [Running tests](#running-tests)
+- [Debugging issues](#debugging-issues)
+  - [Print debugging](#print-debugging)
+  - [Debugging using LLDB](#debugging-using-lldb)
+- [Next steps](#next-steps)
+
+## System Requirements
+
+1. Operating system:
+   The supported operating systems for developing the Swift toolchain are:
+   macOS, Ubuntu Linux LTS, and the latest Ubuntu Linux release.
+   At the moment, Windows is not supported as a host development operating
+   system. Experimental instructions for Windows are available under
+   [Windows.md](/docs/Windows.md).
+2. Python 3: Several utility scripts are written in Python.
+3. Disk space:
+   Make sure that you have enough available disk space before starting.
+   The source code, including full git history, requires about 3.5 GB.
+   Build artifacts take anywhere between 5 GB to 70 GB, depending on the
+   build settings.
+4. Time:
+   Depending on your machine and build settings,
+   a from-scratch build can take a few minutes to several hours,
+   so you might want to grab a beverage while you follow the instructions.
+   Incremental builds are much faster.
+
+## Cloning the project
+
+1. Create a directory for the whole project:
+   ```sh
+   mkdir -p swift-project/swift
+   cd swift-project/swift
+   ```
+2. Clone the sources:
+   - Via SSH (recommended):
+     If you plan on contributing regularly, cloning over SSH provides a better
+     experience. After you've [uploaded your SSH keys to GitHub][]:
```

Comment:
why is ssh recommended? what experiences are better?

Reply:
IIRC you need to repeatedly enter your GitHub password when using HTTPS to clone the different repositories and I also think you need to enter it when pushing changes. Is the suggestion that I explicitly spell out the benefit of using SSH over HTTPS?

Reply:
I use https and the GitHub token gets saved in my login keychain. I only ever enter it once for all of GitHub, no additional use for clones or pushes. For macOS, I think this is as convenient or more convenient than ssh.


In the past, GitHub staff have told me https gets better performance.

I didn't have a suggestion, but now I would suggest removing the recommendation and letting people pick.

Reply:
I ended up looking at this a bit more and turns out:

1. [GitHub recommends using HTTPS over SSH](https://docs.github.com/en/github/using-git/which-remote-url-should-i-use#cloning-with-https-urls-recommended).
2. Credentials aren't required for cloning public repositories but are required for pushing, which may get annoying... (tested this on Linux).

I'm kinda' in two minds on what to do about this. My general tendency when writing documentation for a beginner (in this case, someone beginning to contribute) to:
1. If one option "just works", don't suggest multiple options. For example, I suggest using sccache instead of having a paragraph dedicated to "you could use sccache or you could use cacche, here's how you setup either, here's why you might want one or the other etc."
2. If there is a good reason to suggest multiple options, recommend one option, and briefly describe why they would prefer one over the other.

This is intended to reduce the burden of decision-making; generally the person who is getting started is in a worse position to make that decision than the person writing the documentation, so we make the decision on their behalf.

---

Maybe I should put HTTPS first with a "(recommended by GitHub)" and point out that you may need to enter credentials when pushing changes, which can be avoided using SSH?

Reply:
> My general tendency when writing documentation for a beginner (in this case, someone beginning to contribute)

beginner to contribute to GitHub in general, or Swift in specific? If GitHub in general, maybe there's some existing documentation that compares the two alternatives, and then we can point them there. If focused on a begging contributor to Swift, we sidestep and say to use whichever protocol you typically use.

Reply:
It could be both. Particularly thinking about students here. Generally, I've seen students have some familiarity with GitHub, but that's not always the case. That's why I've tried to be explicit about the git/GitHub-related steps instead of assuming that someone already knows the typical GitHub workflow.

Also, someone experienced might be used to workflows with other tools like Phabricator or Perforce, not so much with GitHub.